README.md
# bip32
[![Build Status](https://travis-ci.org/bitcoinjs/bip32.png?branch=master)](https://travis-ci.org/bitcoinjs/bip32)
[![NPM](https://img.shields.io/npm/v/bip32.svg)](https://www.npmjs.org/package/bip32)

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)


## Example
``` javascript
let bip32 = require('bip32')
let node =  bip32.fromBase58('xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi')

let child = node.derivePath('m/0/0')
// ...
```

#### bip32.Account
``` javascript
let bitcoin = require('bitcoinjs-lib')
let bip32 = require('bip32')
let bip39 = require('bip39')

// ...

let mnemonic = bip39.generateMnemonic()
let seed = bip39.mnemonicToSeed(mnemonic)
let account = bip32.Account.standardFromSeed(seed)

console.log(account.getChainAddress(0))
// => 1QEj2WQD9vxTzsGEvnmLpvzeLVrpzyKkGt

account.nextChainAddress(0)

console.log(account.getChainAddress(0))
// => 1DAi282VN7Ack9o5BqWYkiEsS8Vgx1rLn

console.log(account.getChainAddress(1))
// => 1CXKM323V3kkrHmZQYPUTftGh9VrAWuAYX

console.log(account.derive('1QEj2WQD9vxTzsGEvnmLpvzeLVrpzyKkGt'))
// => xpub6A5Fz4JZg4kd8pLTTaMBKsvVgzRBrvai6ChoxWNTtYQ3UDVG1VyAWQqi6SNqkpsfsx9F8pRqwtKUbU4j4gqpuN2gpgQs4DiJxsJQvTjdzfA

// NOTE: passing in the parent nodes allows for private key escalation (see xprv vs xpub)

console.log(account.derive('1QEj2WQD9vxTzsGEvnmLpvzeLVrpzyKkGt', [external, internal]))
// => xprv9vodQPEygdPGUWeKUVNd6M2N533PvEYP21tYxznauyhrYBBCmdKxRJzmnsTsSNqfTJPrDF98GbLCm6xRnjceZ238Qkf5GQGHk79CrFqtG4d
```

##### BIP44
```
// ...

// equivalent to /44'/0'/0'
let account = bip32.Account.bip44FromSeed(seed, 44, 0, 0)

// ...
```


#### bip32.Chain
``` javascript
let bitcoin = require('bitcoinjs-lib')
let bip32 = require('bip32')

// ...

let hdNode = bitcoin.HDNode.fromSeedHex(seedHex)
let chain = new bip32.Chain(hdNode)

for (let k = 0; k < 10; ++k) chain.next()

let address = chain.get()

console.log(chain.find(address))
// => 9

console.log(chain.pop())
// => address
```


#### bip32.discover
``` javascript
```

## LICENSE [ISC](LICENSE)


CHANGELOG.md
# 0.11.0
* changes `discoverChain` to use a map lookup rather than array ieration

# 0.10.0

# 0.9.0
* add `Account.prototype.derive(address, parents)`
* removes `Account.prototype.getChildrenMap`, replaced with `.derive`

# 0.8.3
* fixes `Account.prototype.discoverChain` mutating the internal **during** discovery (instead of mutating after)

# 0.8.2
* adds `Chain.prototype.derive` with escalation parameter

# 0.8.0
* adds `Account.*JSON` functions
* changes `Account.getChildren` to `Account.getChildrenMap`, returns an address-indexed map and does not throw if not fuond

# 0.7.3
* adds `Chain.clone`

# 0.7.0
* updated peerDependency `bitcoinjs-lib` to `^2.0.0`

# 0.6.1
* adds `Chain.getChains`

# 0.6.0
* adds `Chain.getParent`, `new Chain(...).node` is now private
* changes `Chain` constructor to be lazy
* changes `Chain.indexOf` to `Chain.find`, returns undefined if not found

* adds `Account.getChain`
* removes `Account *(Internal|External)*` APIs, replaced with `*Chain*` API with chain index parameter
* changes `Account` constructor to only accept an array of `Chain`s

# 0.5.0
* removes `Account.getNode`
* changes `Account.getNodes` to `Account.getChildren`
* changes `AddressIterator` to `Chain`


test/chain.js
let bitcoin = require('bitcoinjs-lib')
let Chain = require('../chain')
let test = require('tape')
let fixtures = require('./fixtures/chain')
let createKeccakHash = require('keccak')

function segwitAddr (node) {
  let hash = bitcoin.crypto.hash160(node.getPublicKeyBuffer())
  let script = bitcoin.script.witnessPubKeyHash.output.encode(hash)
  return bitcoin.address.fromOutputScript(script)
}

function ethAddr (node) {
  return '0x' + createKeccakHash('keccak256')
    .update(node.getPublicKeyBuffer())
    .digest().toString('hex')
}

let AF_MAPPING = {
  'eth': ethAddr,
  'segwit': segwitAddr
}

fixtures.forEach(function (f) {
  let node = bitcoin.HDNode.fromBase58(f.node)
  let addressFunction = AF_MAPPING[f.addressFunction]

  test('constructor', function (t) {
    let chain = new Chain(node, null, addressFunction)

    t.plan(3)
    t.equal(chain.k, 0, 'defaults to k=0')

    chain = new Chain(node, f.k, addressFunction)
    t.equal(chain.k, f.k, 'can start at a custom k value')

    chain = new Chain(node, f.k, addressFunction)
    t.equal(chain.addresses.length, 0, 'is lazy')
  })

  test('clone', function (t) {
    let chain = new Chain(node, null, addressFunction)
    let clone = chain.clone()

    t.plan(17)

    // by reference
    t.equal(chain.__parent, clone.__parent, 'same parent')

    // by value
    t.equal(chain.k, clone.k, 'k copied')
    t.notEqual(chain.addresses, clone.addresses, 'address arrays are different objects')
    t.same(chain.addresses, clone.addresses, 'address arrays are deep copied')
    t.notEqual(chain.map, clone.map, 'k maps are different objects')
    t.same(chain.map, clone.map, 'k map is deep copied')
    t.same(chain.addressFunction, clone.addressFunction, 'address function is copied')

    for (let i = 0; i < 7; ++i) chain.next()
    t.equal(clone.k, 0, 'k unchanged by mutation')
    t.notEqual(chain.k, clone.k, 'k unchanged by mutation (2)')
    t.notSame(chain.addresses, clone.addresses, 'address array unchanged by mutation')
    t.notSame(chain.map, clone.map, 'k map unchanged by mutation')

    // re-clone
    clone = chain.clone()

    // re-verify
    t.equal(chain.k, clone.k, 'k copied')
    t.notEqual(chain.addresses, clone.addresses, 'address arrays are different objects')
    t.same(chain.addresses, clone.addresses, 'address arrays are deep copied')
    t.notEqual(chain.map, clone.map, 'k maps are different objects')
    t.same(chain.map, clone.map, 'k map is deep copied')
    t.same(chain.addressFunction, clone.addressFunction, 'address function is copied')
  })

  test('find/derive', function (t) {
    let neutered = node.neutered()
    let chain = new Chain(neutered, f.k, addressFunction)

    t.plan(30)
    for (let i = 0; i < 10; ++i) {
      let address = chain.get()

      t.equal(chain.find(address), f.k + i)
      t.same(chain.derive(address), neutered.derive(f.k + i))
      t.same(chain.derive(address, node), node.derive(f.k + i))
      chain.next()
    }
  })

  test('get', function (t) {
    let chain = new Chain(node, f.k, addressFunction)
    chain.addresses = f.addresses

    t.plan(1)
    t.equal(chain.get(), f.addresses[f.addresses.length - 1], 'returns the last address')
  })

  test('next', function (t) {
    let chain = new Chain(node, f.k - f.addresses.length + 1, addressFunction)

    t.plan(f.addresses.length * 2 + 1)
    f.addresses.forEach(function (x, i) {
      t.equal(chain.get(), f.addresses[i], 'get returns the current address')

      if (f.addresses[i + 1]) {
        t.equal(chain.next(), f.addresses[i + 1], 'next returns the next address')
      }
    })

    t.equal(chain.k, f.k, 'results in the the expected k value')
    t.same(chain.map, f.map, 'results in the expected address map')
  })

  test('pop', function (t) {
    let chain = new Chain(node, null, addressFunction)
    chain.next()
    chain.next()
    let addresses = chain.getAll().concat()

    t.plan(8)
    t.equal(chain.getAll().length, 3, 'has 3 addresses')
    t.equal(chain.get(), addresses[2], 'matches the last address')
    chain.pop()
    t.equal(chain.getAll().length, 2, 'is now 1 less')
    t.equal(chain.get(), addresses[1], 'after pop, matches the second last')
    t.equal(chain.pop(), addresses[1], 'returns the popped address')
    t.equal(chain.getAll().length, 1, 'is now 1 less')
    t.equal(chain.pop(), addresses[0], 'returns the popped address')
    t.equal(chain.pop(), undefined, 'returns undefined when empty')
  })
})


tests/index.js
let BIP32 = require('../')
let tape = require('tape')
let fixtures = require('./fixtures')
let LITECOIN = {
  wif: 0xb0,
  bip32: {
    public: 0x019da462,
    private: 0x019d9cfe
  }
}

let validAll = []
fixtures.valid.forEach(function (f) {
  f.master.network = f.network
  f.master.children = f.children
  f.master.comment = f.comment
  f.children.forEach(function (fc) {
    fc.network = f.network
    validAll.push(fc)
  })
  validAll.push(f.master)
})

function verify (t, hd, prv, f, network) {
  t.equal(hd.chainCode.toString('hex'), f.chainCode)
  t.equal(hd.depth, f.depth >>> 0)
  t.equal(hd.index, f.index >>> 0)
  t.equal(hd.getFingerprint().toString('hex'), f.fingerprint)
  t.equal(hd.getIdentifier().toString('hex'), f.identifier)
  t.equal(hd.getPublicKey().toString('hex'), f.pubKey)
  if (prv) t.equal(hd.toBase58(), f.base58Priv)
  if (prv) t.equal(hd.toWIF(), f.wif)
  if (!prv) t.throws(function () { hd.toWIF() }, /Missing private key/)
  if (!prv) t.equal(hd.d, null) // internal
  t.equal(hd.neutered().toBase58(), f.base58)
  t.equal(hd.isNeutered(), !prv)

  if (!f.children) return
  if (!prv && f.children.some(x => x.hardened)) return

  // test deriving path from master
  f.children.forEach((cf) => {
    let chd = hd.derivePath(cf.path)
    verify(t, chd, prv, cf, network)

    let chdNoM = hd.derivePath(cf.path.slice(2)) // no m/
    verify(t, chdNoM, prv, cf, network)
  })

  // test deriving path from successive children
  let shd = hd
  f.children.forEach((cf) => {
    if (cf.m === undefined) return
    if (cf.hardened) {
      shd = shd.deriveHardened(cf.m)
    } else {
      // verify any publicly derived children
      if (cf.base58) verify(t, shd.neutered().derive(cf.m), false, cf, network)

      shd = shd.derive(cf.m)
      verify(t, shd, prv, cf, network)
    }

    t.throws(function () {
      shd.derivePath('m/0')
    }, /Expected master, got child/)

    verify(t, shd, prv, cf, network)
  })
}

validAll.forEach(function (ff) {
  tape.test(ff.comment || ff.base58Priv, function (t) {
    let network
    if (ff.network === 'litecoin') network = LITECOIN

    let hd = BIP32.fromBase58(ff.base58Priv, network)
    verify(t, hd, true, ff, network)

    hd = BIP32.fromBase58(ff.base58, network)
    verify(t, hd, false, ff, network)

    if (ff.seed) {
      hd = BIP32.fromSeed(Buffer.from(ff.seed, 'hex'), network)
      verify(t, hd, true, ff, network)
    }

    t.end()
  })
})

tape.test('fromBase58 throws', function (t) {
  fixtures.invalid.fromBase58.forEach(function (f) {
    t.throws(function () {
      let network
      if (f.network === 'litecoin') network = LITECOIN

      BIP32.fromBase58(f.string, network)
    }, new RegExp(f.exception))
  })

  t.end()
})

tape.test('works for Private -> public (neutered)', function (t) {
  let f = fixtures.valid[1]
  let c = f.children[0]

  let master = BIP32.fromBase58(f.master.base58Priv)
  let child = master.derive(c.m).neutered()

  t.plan(1)
  t.equal(child.toBase58(), c.base58)
})

tape.test('works for Private -> public (neutered, hardened)', function (t) {
  let f = fixtures.valid[0]
  let c = f.children[0]

  let master = BIP32.fromBase58(f.master.base58Priv)
  let child = master.deriveHardened(c.m).neutered()

  t.plan(1)
  t.equal(c.base58, child.toBase58())
})

tape.test('works for Public -> public', function (t) {
  let f = fixtures.valid[1]
  let c = f.children[0]

  let master = BIP32.fromBase58(f.master.base58)
  let child = master.derive(c.m)

  t.plan(1)
  t.equal(c.base58, child.toBase58())
})

tape.test('throws on Public -> public (hardened)', function (t) {
  let f = fixtures.valid[0]
  let c = f.children[0]

  let master = BIP32.fromBase58(f.master.base58)

  t.plan(1)
  t.throws(function () {
    master.deriveHardened(c.m)
  }, /Missing private key for hardened child key/)
})

tape.test('throws on wrong types', function (t) {
  let f = fixtures.valid[0]
  let master = BIP32.fromBase58(f.master.base58)

  fixtures.invalid.derive.forEach(function (fx) {
    t.throws(function () {
      master.derive(fx)
    }, /Expected UInt32/)
  })

  fixtures.invalid.deriveHardened.forEach(function (fx) {
    t.throws(function () {
      master.deriveHardened(fx)
    }, /Expected UInt31/)
  })

  fixtures.invalid.derivePath.forEach(function (fx) {
    t.throws(function () {
      master.derivePath(fx)
    }, /Expected BIP32Path, got/)
  })

  t.end()
})

tape.test('works when private key has leading zeros', function (t) {
  let key = 'xprv9s21ZrQH143K3ckY9DgU79uMTJkQRLdbCCVDh81SnxTgPzLLGax6uHeBULTtaEtcAvKjXfT7ZWtHzKjTpujMkUd9dDb8msDeAfnJxrgAYhr'
  let hdkey = BIP32.fromBase58(key)

  t.plan(2)
  t.equal(hdkey.d.toString('hex'), '00000055378cf5fafb56c711c674143f9b0ee82ab0ba2924f19b64f5ae7cdbfd')
  let child = hdkey.derivePath('m/44\'/0\'/0\'/0/0\'')
  t.equal(child.d.toString('hex'), '3348069561d2a0fb925e74bf198762acc47dce7db27372257d2d959a9e6f8aeb')
})

/*
tape.test('fromSeed', function (t) {
  t.throws(function () {
    this.mock(BigInteger).expects('fromBuffer')
      .once().returns(BigInteger.ZERO)

    t.throws(function () {
      BIP32.fromSeed('ffffffffffffffffffffffffffffffff')
    }, /Private key must be greater than 0/)
  }, /ErrorXXX/, 'throws if IL is not within interval [1, n - 1] | IL === 0')

'throws if IL is not within interval [1, n - 1] | IL === n'
    this.mock(BigInteger).expects('fromBuffer')
      .once().returns(curve.n)

    t.throws(function () {
      BIP32.fromSeedHex('ffffffffffffffffffffffffffffffff')
    }, /Private key must be less than the curve order/)

'throws on low entropy seed'
    t.throws(function () {
      BIP32.fromSeedHex('ffffffffff')
    }, /Seed should be at least 128 bits/)

'throws on too high entropy seed'
    t.throws(function () {
      BIP32.fromSeedHex('ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff')
    }, /Seed should be at most 512 bits/)

})

//  let hd = BIP32.fromSeed(Buffer.alloc(64))
//
//  tape.test('sign', function () {
//    this.mock(keyPair).expects('sign')
//      .once().withArgs(hash).returns('signed')
//
//    t.equal(hd.sign(hash), 'signed')
//  })
//
//  tape.test('verify', function (t) {
//    let signature = hd.sign(hash)
//
//    this.mock(keyPair).expects('verify')
//      .once().withArgs(hash, signature).returns('verified')
//
//    t.equal(hd.verify(hash, signature), 'verified')
//  })
*/


test/account.js
let Account = require('../account')
let Chain = require('../chain')
let test = require('tape')

let f = require('./fixtures/account')
f.allAddresses = [].concat.apply([], f.addresses)

function blankAccount (json) {
  let account = Account.fromJSON(json)
  let chains = account.chains.map(function (chain) {
    return new Chain(chain.__parent, 0)
  })

  return new Account(chains)
}

test('containsAddress', function (t) {
  let account = Account.fromJSON(f.neutered.json)

  f.allAddresses.forEach(function (address) {
    t.equal(account.containsAddress(address), true, 'returns true for known chain address')
  })

  t.equal(account.containsAddress('mpFZW4A9QtRuSpuh9SmeW7RSzFE3TgB8Ko'), false, 'returns false for unknown address')
  t.end()
})

test('clone', function (t) {
  let account = Account.fromJSON(f.neutered.json)
  let clone = account.clone()

  // by reference
  t.equal(account.chains.length, clone.chains.length, 'same number of chains')
  t.notEqual(account.chains, clone.chains, 'chain arrays are different arrays')
  t.same(account.chains, clone.chains, 'chains are deep copied')
  for (let i = 0; i < account.chains.length; ++i) {
    t.notEqual(account.chains[i], clone.chains[i], 'chains are different objects')
  }

  account.nextChainAddress(1)
  t.notSame(account.chains, clone.chains, 'chains now diverge, 1 has an extra address')
  t.end()
})

test('bip44', function (t) {
  let seed = Buffer.alloc(32, 0x03)
  let account = Account.bip44FromSeed(seed, 44, 0x00, 0)

  t.same(account.getChainAddress(0), '')
  t.same(account.getChainAddress(1), '')
  t.end()
})

test('discoverChain', function (t) {
  let account = Account.fromJSON(f.neutered.json)
  let before = account.getChainAddress(0)
  let after = account.getChain(0).clone().next()

  t.test('does not mutate the chain during discovery', function (t) {
    t.plan(2)

    account.discoverChain(0, 20, function (addresses, callback) {
      return callback(null, addresses.map(function (address) {
        // account.containsAddress would return true if internally the chain was iterating
        return address !== before && account.containsAddress(address)
      }))
    }, function (err) {
      t.ifErr(err, 'no error')
      t.equal(account.getChainAddress(0), before, 'internal chain was unchanged')
    })
  })

  t.test('does mutate the chain post-discovery', function (t) {
    t.plan(2)

    account.discoverChain(0, 20, function (addresses, callback) {
      return callback(null, addresses.map(function (address) {
        return account.containsAddress(address)
      }))
    }, function (err) {
      t.ifErr(err, 'no error')
      t.equal(account.getChainAddress(0), after, 'internal chain iterated forward one address')
    })
  })
})

test('getAllAddresses', function (t) {
  let account = blankAccount(f.neutered.json)

  t.plan(2)
  t.equal(account.getAllAddresses().length, 2, 'returns only 2 addresses post-construction')

  // iterate the chains
  f.addresses.forEach(function (a, i) {
    for (let j = 1; j < a.length; ++j) account.nextChainAddress(i)
  })

  t.same(account.getAllAddresses(), f.allAddresses, 'returns all derived addresses')
})

test('getChainAddress', function (t) {
  let account = blankAccount(f.neutered.json)

  f.addresses.forEach(function (addresses, i) {
    addresses.forEach(function (address) {
      t.equal(account.getChainAddress(i), address, 'matches the latest chain address')
      account.nextChainAddress(i)
    })
  })

  t.end()
})

test('getNetwork', function (t) {
  let account = Account.fromJSON(f.neutered.json)

  t.plan(1)
  t.equal(account.getNetwork(), account.chains[0].__parent.keyPair.network, 'matches keyPair network')
})

test('isChainAddress', function (t) {
  let account = Account.fromJSON(f.neutered.json)

  f.addresses.forEach(function (addresses, i) {
    addresses.forEach(function (address) {
      t.equal(account.isChainAddress(i, address), true, 'for same chain')
      t.equal(account.isChainAddress(i === 1 ? 0 : 1, address), false, 'for different chains')
    })
  })

  t.end()
})

test('nextChainAddress', function (t) {
  let account = blankAccount(f.neutered.json)

  // returns the new address
  f.addresses.forEach(function (addresses, i) {
    // skip the first address
    addresses.slice(1).forEach(function (address, j) {
      t.equal(account.getChainAddress(i), addresses[j], 'is moving forward the chain')
      t.equal(account.nextChainAddress(i), address, 'returns the next address: ' + address)
    })
  })

  t.end()
})

test('getChain', function (t) {
  let account = blankAccount(f.neutered.json)

  f.neutered.json.forEach(function (_, i) {
    t.equal(typeof account.getChain(i), 'object')
    t.equal(account.getChain(i), account.chains[i], 'matches internal .chain')
  })

  t.end()
})

test('getChains', function (t) {
  let account = blankAccount(f.neutered.json)

  t.plan(2)
  t.equal(account.getChains().length, f.neutered.json.length, 'returns the expected number of chains')
  t.equal(account.getChains(), account.chains, 'matches internal .chains')
})

test('derive', function (t) {
  let neutered = Account.fromJSON(f.neutered.json)

  t.test('neutered node', function (t) {
    f.addresses.forEach(function (addresses, i) {
      addresses.forEach(function (address, j) {
        let actual = neutered.derive(address)
        let expected = f.neutered.children[i][j]

        t.equal(expected, actual.toBase58(), 'return a neutered node')
      })
    })

    let unknown = neutered.derive('mpFZW4A9QtRuSpuh9SmeW7RSzFE3TgB8Ko')
    t.equal(undefined, unknown, 'ignores unknown addresses')
    t.end()
  })

  let priv = Account.fromJSON(f.private.json)

  t.test('neutered node w/ escalation', function (t) {
    let privParents = priv.chains.map(function (x) { return x.__parent })

    f.addresses.forEach(function (addresses, i) {
      addresses.forEach(function (address, j) {
        let actual = neutered.derive(address, privParents)
        let expected = f.private.children[i][j]

        t.equal(expected, actual.toBase58(), 'returns a private node')
      })
    })

    let unknown = neutered.derive('mpFZW4A9QtRuSpuh9SmeW7RSzFE3TgB8Ko')
    t.equal(undefined, unknown, 'ignores unknown addresses')
    t.end()
  })

  t.test('private node', function (t) {
    f.addresses.forEach(function (addresses, i) {
      addresses.forEach(function (address, j) {
        let actual = priv.derive(address)
        let expected = f.private.children[i][j]

        t.equal(expected, actual.toBase58(), 'returns a private node')
      })
    })

    let unknown = neutered.derive('mpFZW4A9QtRuSpuh9SmeW7RSzFE3TgB8Ko')
    t.equal(undefined, unknown, 'ignores unknown addresses')
    t.end()
  })

  t.end()
})

// TODO
test('discoverChain', function (t) {
  // .getChainAddress() should remain the same after a uneventful discovery
  // .getChainAddress() should change after an eventful discovery
  t.end()
})

test('toJSON', function (t) {
  let neutered = Account.fromJSON(f.neutered.json)
  let priv = Account.fromJSON(f.private.json)

  t.plan(2)
  t.same(neutered.toJSON(), f.neutered.json, 'neutered json matches fixtures')
  t.same(priv.toJSON(), f.private.json, 'priv json matches fixtures')
})


ecc.js
let secp256k1 = require('secp256k1')

function intAdd (a, b) {
  return secp256k1.privateKeyTweakAdd(a, b)
}

var EC_ZERO = Buffer.from('0000000000000000000000000000000000000000000000000000000000000000', 'hex')
var EC_UINT_MAX = Buffer.from('fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141', 'hex')

function intIsZero (a) {
  return a.equals(EC_ZERO)
}

function intVerify (value) {
//    return secp256k1.privateKeyVerify(a)
  return Buffer.isBuffer(value) &&
    value.length === 32 &&
    value.compare(EC_ZERO) > 0 && // > 0
    value.compare(EC_UINT_MAX) < 0 // < n-1
}

function pointAddTweak (q, tweak, compressed) {
  try {
    return secp256k1.publicKeyTweakAdd(q, tweak, compressed)
  } catch (e) {
    return null
  }
}

function pointDerive (d, compressed) {
  return secp256k1.publicKeyCreate(d, compressed)
}

function pointVerify (q, compressed) {
  if (!Buffer.isBuffer(q)) return false
  if (q.length < 33) return false

  switch (q[0]) {
    case 0x02:
    case 0x03:
      if (q.length !== 33) return false
      break
    case 0x04:
      if (q.length !== 65) return false
      if (compressed) return false
      break
    default:
      return false
  }

  return secp256k1.publicKeyVerify(q)
}

function ecdsaSign (hash, d) {
  return secp256k1.sign(hash, d).signature
}

function ecdsaVerify (hash, signature, Q) {
  return secp256k1.verify(hash, signature, Q)
}

module.exports = {
  intAdd,
  intIsZero,
  intVerify,
  pointAddTweak,
  pointDerive,
  pointVerify,
  ecdsaSign,
  ecdsaVerify
}


index.js
let Buffer = require('safe-buffer').Buffer
let base58check = require('bs58check')
let crypto = require('./crypto')
let ecc = require('./ecc')
let typeforce = require('typeforce')
let wif = require('wif')

let NETWORK_TYPE = typeforce.compile({
  wif: typeforce.UInt8,
  bip32: {
    public: typeforce.UInt32,
    private: typeforce.UInt32
  }
})

let BITCOIN = {
  wif: 0x80,
  bip32: {
    public: 0x0488b21e,
    private: 0x0488ade4
  }
}

function BIP32 (d, Q, chainCode, network) {
  this.d = d || null
  this.Q = Q || ecc.pointDerive(d, true)
  if (this.Q === null) throw new TypeError('Bad public key')

  this.chainCode = chainCode
  this.depth = 0
  this.index = 0
  this.network = network
  this.parentFingerprint = 0x00000000
}

function fromSeed (seed, network) {
  typeforce(typeforce.Buffer, seed)
  if (seed.length < 16) throw new TypeError('Seed should be at least 128 bits')
  if (seed.length > 64) throw new TypeError('Seed should be at most 512 bits')
  if (network) typeforce(NETWORK_TYPE, network)
  network = network || BITCOIN

  let I = crypto.hmacSHA512('Bitcoin seed', seed)
  let IL = I.slice(0, 32)
  let IR = I.slice(32)

  // if IL is 0 or >= n, the master key is invalid
  if (!ecc.isPrivateKey(IL)) throw new TypeError('Private key not in range [1, n)')

  return new BIP32(IL, null, IR, network)
}

function fromBase58 (string, network) {
  let buffer = base58check.decode(string)
  if (buffer.length !== 78) throw new TypeError('Invalid buffer length')
  if (network) typeforce(NETWORK_TYPE, network)
  network = network || BITCOIN

  // 4 bytes: version bytes
  let version = buffer.readUInt32BE(0)
  if (version !== network.bip32.private &&
    version !== network.bip32.public) throw new TypeError('Invalid network version')

  // 1 byte: depth: 0x00 for master nodes, 0x01 for level-1 descendants, ...
  let depth = buffer[4]

  // 4 bytes: the fingerprint of the parent's key (0x00000000 if master key)
  let parentFingerprint = buffer.readUInt32BE(5)
  if (depth === 0) {
    if (parentFingerprint !== 0x00000000) throw new TypeError('Invalid parent fingerprint')
  }

  // 4 bytes: child number. This is the number i in xi = xpar/i, with xi the key being serialized.
  // This is encoded in MSB order. (0x00000000 if master key)
  let index = buffer.readUInt32BE(9)
  if (depth === 0 && index !== 0) throw new TypeError('Invalid index')

  // 32 bytes: the chain code
  let chainCode = buffer.slice(13, 45)
  let hd

  // 33 bytes: private key data (0x00 + k)
  if (version === network.bip32.private) {
    if (buffer.readUInt8(45) !== 0x00) throw new TypeError('Invalid private key')
    let k = buffer.slice(46, 78)

    // if IL is 0 or >= n, the master key is invalid
    if (!ecc.isPrivateKey(k)) throw new TypeError('Private key not in range [1, n)')
    hd = new BIP32(k, null, chainCode, network)

  // 33 bytes: public key data (0x02 + X or 0x03 + X)
  } else {
    let X = buffer.slice(45, 78)

    // verify the X coordinate is a point on the curve
    if (!ecc.isPublicKey(X)) throw new TypeError('Point is not on the curve')
    hd = new BIP32(null, X, chainCode, network)
  }

  hd.depth = depth
  hd.index = index
  hd.parentFingerprint = parentFingerprint
  return hd
}

BIP32.prototype.getIdentifier = function () {
  return crypto.hash160(this.Q)
}

BIP32.prototype.getFingerprint = function () {
  return this.getIdentifier().slice(0, 4)
}

BIP32.prototype.getNetwork = function () {
  return this.network
}

BIP32.prototype.getPublicKeyBuffer = function () {
  return this.Q
}

BIP32.prototype.neutered = function () {
  let neutered = new BIP32(null, this.Q, this.chainCode, this.network)
  neutered.depth = this.depth
  neutered.index = this.index
  neutered.parentFingerprint = this.parentFingerprint
  return neutered
}

BIP32.prototype.toBase58 = function () {
  let network = this.network
  let version = (!this.isNeutered()) ? network.bip32.private : network.bip32.public
  let buffer = Buffer.allocUnsafe(78)

  // 4 bytes: version bytes
  buffer.writeUInt32BE(version, 0)

  // 1 byte: depth: 0x00 for master nodes, 0x01 for level-1 descendants, ....
  buffer.writeUInt8(this.depth, 4)

  // 4 bytes: the fingerprint of the parent's key (0x00000000 if master key)
  buffer.writeUInt32BE(this.parentFingerprint, 5)

  // 4 bytes: child number. This is the number i in xi = xpar/i, with xi the key being serialized.
  // This is encoded in big endian. (0x00000000 if master key)
  buffer.writeUInt32BE(this.index, 9)

  // 32 bytes: the chain code
  this.chainCode.copy(buffer, 13)

  // 33 bytes: the public key or private key data
  if (!this.isNeutered()) {
    // 0x00 + k for private keys
    buffer.writeUInt8(0, 45)
    this.d.copy(buffer, 46)

  // 33 bytes: the public key
  } else {
    // X9.62 encoding for public keys
    this.Q.copy(buffer, 45)
  }

  return base58check.encode(buffer)
}

BIP32.prototype.toWIF = function () {
  if (!this.d) throw new TypeError('Missing private key')
  return wif.encode(this.network.wif, this.d, true)
}

let HIGHEST_BIT = 0x80000000

// https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#child-key-derivation-ckd-functions
BIP32.prototype.derive = function (index) {
  typeforce(typeforce.UInt32, index)

  let isHardened = index >= HIGHEST_BIT
  let data = Buffer.allocUnsafe(37)

  // Hardened child
  if (isHardened) {
    if (this.isNeutered()) throw new TypeError('Missing private key for hardened child key')

    // data = 0x00 || ser256(kpar) || ser32(index)
    data[0] = 0x00
    this.d.copy(data, 1)
    data.writeUInt32BE(index, 33)

  // Normal child
  } else {
    // data = serP(point(kpar)) || ser32(index)
    //      = serP(Kpar) || ser32(index)
    this.Q.copy(data, 0)
    data.writeUInt32BE(index, 33)
  }

  let I = crypto.hmacSHA512(this.chainCode, data)
  let IL = I.slice(0, 32)
  let IR = I.slice(32)

  // if parse256(IL) >= n, proceed with the next value for i
  if (!ecc.isPrivateKey(IL)) return this.derive(index + 1)

  // Private parent key -> private child key
  let hd
  if (!this.isNeutered()) {
    // ki = parse256(IL) + kpar (mod n)
    let ki = ecc.privateAdd(this.d, IL)

    // In case ki == 0, proceed with the next value for i
    if (ki == null) return this.derive(index + 1)

    hd = new BIP32(ki, null, IR, this.network)

  // Public parent key -> public child key
  } else {
    // Ki = point(parse256(IL)) + Kpar
    //    = G*IL + Kpar
    let Ki = ecc.publicAdd(this.Q, IL, true)

    // In case Ki is the point at infinity, proceed with the next value for i
    if (Ki === null) return this.derive(index + 1)

    hd = new BIP32(null, Ki, IR, this.network)
  }

  hd.depth = this.depth + 1
  hd.index = index
  hd.parentFingerprint = this.getFingerprint().readUInt32BE(0)
  return hd
}

let UINT31_MAX = Math.pow(2, 31) - 1
function UInt31 (value) {
  return typeforce.UInt32(value) && value <= UINT31_MAX
}

BIP32.prototype.deriveHardened = function (index) {
  typeforce(UInt31, index)

  // Only derives hardened private keys by default
  return this.derive(index + HIGHEST_BIT)
}

// Private === not neutered
// Public === neutered
BIP32.prototype.isNeutered = function () {
  return this.d === null
}

function isBIP32Path (value) {
  return typeforce.String(value) && value.match(/^(m\/)?(\d+'?\/)*\d+'?$/)
}

BIP32.prototype.derivePath = function (path) {
  typeforce(isBIP32Path, path)

  let splitPath = path.split('/')
  if (splitPath[0] === 'm') {
    if (this.parentFingerprint) throw new TypeError('Expected master, got child')

    splitPath = splitPath.slice(1)
  }

  return splitPath.reduce(function (prevHd, indexStr) {
    let index
    if (indexStr.slice(-1) === "'") {
      index = parseInt(indexStr.slice(0, -1), 10)
      return prevHd.deriveHardened(index)
    } else {
      index = parseInt(indexStr, 10)
      return prevHd.derive(index)
    }
  }, this)
}

BIP32.prototype.sign = function (hash) {
  return ecc.sign(hash, this.d)
}

BIP32.prototype.verify = function (hash, signature) {
  return ecc.verify(hash, signature, this.Q)
}

module.exports = {
  fromBase58,
  fromSeed
}
